### PR TITLE
Element before module

### LINKED ISSUE #26: dom-module must come before the script that uses it
Port from polylint. See domModuleAfterPolymer here: https://github.com/PolymerLabs/polylint/blob/master/src/linters.ts#L457

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

Fixes #26
Supercedes #37 - sorry @43081j! This was just the last rule that needed to get ported from polylint. It was surprisingly complicated!

---
This change is [Reviewable](https://reviewable.io/reviews/polymer/polymer-linter/50)

### CODE REVIEW

File: src/polymer/element-before-dom-module.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            })
+            .filter((v) => !!v);
+
+    // Sort the element groups by the order in which they appear in the
+    // document.
+    const sorted = localElements.concat(elementsByImport).sort((a, b) => {
+      return comparePositionAndRange(a.sourceRange.start, b.sourceRange);
+    });
+
+    const seenSoFar = new Set<null|string>([null]);
```
Comment (usergenic) [inline src/polymer/element-before-dom-module.ts]: I can't figure out why seenSoFar needs to start with an array containing a null and/or allow nulls?

**Review by usergenic — APPROVED** [review]
LGTM with a question

File: src/polymer/element-before-dom-module.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      Warns for an element being declared before its dom-module.
+
+      For example, this is invalid:
+        <script>Polymer({is: 'my-elem'})</script>
+        <dom-module id='my-elem'></dom-module>
+
+      But this is fine:
+        <dom-module id='my-elem'></dom-module>
+        <script>Polymer({is: 'my-elem'})</script>
+      `);
```
Comment (usergenic) [inline src/polymer/element-before-dom-module.ts]: You `.trim()` the other descriptions.  Maybe trim this one too.

File: src/polymer/element-before-dom-module.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
```
Comment (valdrinkoshi) [inline src/polymer/element-before-dom-module.ts]: (c) 2017 😬

File: src/polymer/element-before-dom-module.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        seenSoFar.add(tagName);
+
+        // Ok! Finally! domModule is a <dom-module> from `document`, and
+        // `element` is its element definition, first defined at
+        // `pair.sourceRange` in `document`. Now we compare them, and if the
+        // element comes before the dom-module, that's an error!
+        if (comparePositionAndRange(
+                pair.sourceRange.start, domModule.sourceRange) === -1) {
+          // TODO(rictic): if we ever support multiple source ranges on
+          //     warnings, this would be a good canidadate.
```
Comment (valdrinkoshi) [inline src/polymer/element-before-dom-module.ts]: nit: typo for candidate

File: src/polymer/element-before-dom-module.ts (hunk 1 above)
Comment (rictic) [reply src/polymer/element-before-dom-module.ts]: Good catch, from an earlier version of the code where that's how I was dealing with unknown tag names. Fixed.

File: src/polymer/element-before-dom-module.ts (hunk 2 above)
Comment (rictic) [reply src/polymer/element-before-dom-module.ts]: +1, done

File: src/polymer/element-before-dom-module.ts (hunk 3 above)
Comment (rictic) [reply src/polymer/element-before-dom-module.ts]: Arg copy-pasta gets me again.

File: src/polymer/element-before-dom-module.ts (hunk 4 above)
Comment (rictic) [reply src/polymer/element-before-dom-module.ts]: image

**Pull request merged**